### PR TITLE
IDENTITY-5182: Improve Caching configuration in OAuth

### PR DESCRIPTION
**Pull request by mefarazath**
### Proposed changes in this pull request

https://wso2.org/jira/browse/IDENTITY-5182
- Remove the code segments of relying on `<EnableOAuthCache>` config in identity.xml to decide whether cache is on/off in OAuth
-  Remove parsing <EnableOAuthCache> config from identity.xml and deprecate isCacheEnabled() method in OAuthServerConfiguration class.

### Follow up actions
-  We need to remove `<EnableOAuthCache>` config from identity.xml as a POST action.

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labelled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java [hunk 1]
```diff
@@ -1208,17 +1210,17 @@ private void parseOAuthURLs(OMElement oauthConfigElem) {
...
-                oauthConfigElem.getFirstChildWithName(getQNameWithIdentityNS(ConfigElements.ENABLE_CACHE));
-        if (enableCacheElem != null) {
-            cacheEnabled = Boolean.parseBoolean(enableCacheElem.getText());
-        }
-
-        if (log.isDebugEnabled()) {
-            log.debug("Enable OAuth Cache was set to : " + cacheEnabled);
-        }
-    }
+//    private void parseCachingConfiguration(OMElement oauthConfigElem) {
```
Comment (madurangasiriwardena) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Can we remove the commented code?

Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/config/OAuthServerConfiguration.java]: Done.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java [hunk 2]
```diff
@@ -308,15 +308,15 @@ public OAuthRevocationResponseDTO revokeTokenByOAuthClient(OAuthRevocationReques
...
-                        // cache hit, do the type check.
-                        if (result instanceof AccessTokenDO) {
-                            accessTokenDO = (AccessTokenDO) result;
-                        }
+
+                    OAuthCacheKey cacheKey = new OAuthCacheKey(revokeRequestDTO.getToken());
+                    CacheEntry result = OAuthCache.getInstance().getValueFromCache(cacheKey);
+
+                    // check cache hit, do the type check.
+                    if (result!= null && result instanceof AccessTokenDO) {
```
Comment (isharak) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/OAuth2Service.java]: Edit formatting

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java [hunk 3]
```diff
@@ -46,14 +46,14 @@
 
     private Map<String, ResponseTypeHandler> responseHandlers = new HashMap<>();
 
-    private AppInfoCache appInfoCache;
+//    private AppInfoCache appInfoCache;
```
Comment (isharak) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/AuthorizationHandlerManager.java]: remove this commented line

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java [hunk 4]
```diff
@@ -404,20 +400,20 @@ public static String getAuthenticatedUsername(String clientId, String clientSecr
         boolean isUsernameCaseSensitive = IdentityUtil.isUserStoreInUsernameCaseSensitive(username);
 
         if (OAuth2Util.authenticateClient(clientId, clientSecretProvided)) {
-            // check cache
-            if (cacheEnabled) {
```
Comment (isharak) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: deprecate this method

Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: added @depricate tag

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java [hunk 5]
```diff
@@ -427,17 +423,18 @@ public static String getAuthenticatedUsername(String clientId, String clientSecr
...
+            if (username != null && !cacheHit) {
+                /*
+                  Using the same ClientCredentialDO to host username. Semantically wrong since ClientCredentialDo
+                  accept a client secret and we're storing a username in the secret variable. Do we have to make our
+                  own cache key and cache entry class every time we need to put something to it? Ideal solution is
+                  to have a generalized way of caching a key:value pair
                  */
                 if (isUsernameCaseSensitive) {
-                    cache.addToCache(new OAuthCacheKey(clientId + ":" + username), new ClientCredentialDO(username));
+                    OAuthCache.getInstance()
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/util/OAuth2Util.java]: Done!

**Comment by mefarazath** [comment]
@isharak Added requested changes.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/AbstractResponseTypeHandler.java [hunk 6]
```diff
@@ -48,17 +48,17 @@
     protected OauthTokenIssuer oauthIssuerImpl;
     protected TokenMgtDAO tokenMgtDAO;
     protected boolean cacheEnabled;
-    protected OAuthCache oauthCache;
+    protected OAuthCache oAuthCache;
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/authz/handlers/AbstractResponseTypeHandler.java]: API change

**Pull request merged**